File: test/python/payment_gateway/consult_pay_test.py
import os
import pytest
from dana.utils.snap_configuration import SnapConfiguration, AuthSettings, Env
from dana.payment_gateway.v1.enum import *
from dana.payment_gateway.v1.models import *
from dana.payment_gateway.v1.models import ConsultPayRequest
from dana.payment_gateway.v1 import *
from dana.api_client import ApiClient
from dana.exceptions import *

from helper.util import get_request, with_delay
from helper.assertion import *
from helper.api_helpers import execute_and_assert_api_error, get_headers_with_signature

title_case = "ConsultPay"
json_path_file = "resource/request/components/PaymentGateway.json"

configuration = SnapConfiguration(
    api_key=AuthSettings(
        PRIVATE_KEY=os.environ.get("PRIVATE_KEY"),
        ORIGIN=os.environ.get("ORIGIN"),
        X_PARTNER_ID=os.environ.get("X_PARTNER_ID"),
        ENV=Env.SANDBOX,
        DANA_ENV=Env.SANDBOX,
        X_DEBUG="true"
    )
)

with ApiClient(configuration) as api_client:
    api_instance = PaymentGatewayApi(api_client)


@with_delay()
def test_consult_pay_success():
    """Should give success response code and message and correct mandatory fields"""
    case_name = "ConsultPayBalancedSuccess"
    
    # Get the request data from the JSON file
    json_dict = get_request(json_path_file, title_case, case_name)
    
    # Convert the request data to a ConsultPayRequest object
    consult_pay_request_obj = ConsultPayRequest.from_dict(json_dict)
    consult_pay_request_obj.merchant_id = os.environ.get("MERCHANT_ID")
    
    # Make the API call
    api_response = api_instance.consult_pay(consult_pay_request_obj)
    
    # Assert the API response
    assert_response(json_path_file, title_case, case_name, CancelOrderResponse.to_json(api_response), None)


@with_delay()
def test_consult_pay_invalid_field_format():
    """Should give fail response code and message and correct mandatory fields"""
    case_name = "ConsultPayBalancedInvalidFieldFormat"
        
    # Get the request data from the JSON file
    json_dict = get_request(json_path_file, title_case, case_name)
    
    # Convert the request data to a ConsultPayRequest object
    consult_pay_request_obj = ConsultPayRequest.from_dict(json_dict)
    
    # Make the API call
    try:
        api_response = api_instance.consult_pay(consult_pay_request_obj)
    except BadRequestException as e:
        assert_fail_response(json_path_file, title_case, case_name, e.body)
        
        
@with_delay()
def test_consult_pay_invalid_mandatory_field():
    """Should give fail response code and message and correct mandatory fields"""
    case_name = "ConsultPayBalancedInvalidMandatoryField"
        
    # Get the request data from the JSON file
    json_dict = get_request(json_path_file, title_case, case_name)
    
    # Convert the request data to a ConsultPayRequest object
    consult_pay_request_obj = ConsultPayRequest.from_dict(json_dict)
    
    # Prepare headers without timestamp to trigger mandatory field error
    # This now handles the signature generation internally
    headers = get_headers_with_signature(
        method="POST",
        resource_path="/v1.0/payment-gateway/consult-pay.htm",
        request_obj=json_dict,
        with_timestamp=False
    )
    
    # Execute the API request and assert the error
    execute_and_assert_api_error(
        api_client,
        "POST",
        "http://api.sandbox.dana.id/v1.0/payment-gateway/consult-pay.htm",
        consult_pay_request_obj,
        headers,
        400,  # Expected status code
        json_path_file,
        title_case,
        case_name
    )

File: test/php/payment_gateway/ConsultPayTest.php
<?php

namespace DanaUat\PaymentGateway;

use Dana\Configuration;
use Dana\PaymentGateway\v1\Api\PaymentGatewayApi;
use Dana\PaymentGateway\v1\Model\ConsultPayRequest;
use Dana\ObjectSerializer;
use Dana\Env;
use DanaUat\Helper\Util;
use DanaUat\Helper\Assertion;
use Dana\ApiException;
use PHPUnit\Framework\TestCase;
use Exception;

class ConsultPayTest extends TestCase
{
    private static $titleCase = 'ConsultPay';
    private static $jsonPathFile = 'resource/request/components/PaymentGateway.json';
    private static $apiInstance;
    private static $merchantId;

    public static function setUpBeforeClass(): void
    {
        // Set up configuration with authentication settings
        $configuration = new Configuration();
        
        // The Configuration constructor automatically loads values from environment variables
        // But we can manually set them if needed
        $configuration->setApiKey('PRIVATE_KEY', getenv('PRIVATE_KEY'));
        $configuration->setApiKey('ORIGIN', getenv('ORIGIN'));
        $configuration->setApiKey('X_PARTNER_ID', getenv('X_PARTNER_ID'));
        $configuration->setApiKey('ENV', Env::SANDBOX);
        
        // Create PaymentGatewayApi instance directly with configuration
        self::$apiInstance = new PaymentGatewayApi(null, $configuration);
        
        // Store merchantId for reuse
        self::$merchantId = getenv('MERCHANT_ID');
    }

    /**
     * Should give success response code and message and correct mandatory fields
     */
    public function testConsultPaySuccess(): void
    {
        Util::withDelay(function() {
            $caseName = 'ConsultPayBalancedSuccess';
            
            try {
                // Get the request data from the JSON file
                $jsonDict = Util::getRequest(
                    self::$jsonPathFile,
                    self::$titleCase,
                    $caseName
                );
                
                // Set merchant ID
                $jsonDict['merchantId'] = self::$merchantId;
                
                // Create a ConsultPayRequest object from the JSON request data
                $consultPayRequestObj = ObjectSerializer::deserialize(
                    $jsonDict,
                    'Dana\PaymentGateway\v1\Model\ConsultPayRequest'
                );
                
                // Make the API call
                $apiResponse = self::$apiInstance->consultPay($consultPayRequestObj);
                
                // Assert the response matches the expected data
                Assertion::assertResponse(
                    self::$jsonPathFile,
                    self::$titleCase,
                    $caseName,
                    $apiResponse->__toString()
                );
                
                $this->assertTrue(true);
            } catch (ApiException $e) {
                $this->fail('Failed to consult pay: ' . $e->getMessage());
            } catch (Exception $e) {
                $this->fail('Unexpected exception: ' . $e->getMessage());
            }
        });
    }

    /**
     * Should give fail response code and message for invalid field format
     */
    public function testConsultPayInvalidFieldFormat(): void
    {
        Util::withDelay(function() {
            $caseName = 'ConsultPayBalancedInvalidFieldFormat';

            $jsonDict = Util::getRequest(
                    self::$jsonPathFile,
                    self::$titleCase,
                    $caseName
                );

            // Set merchant ID
                $jsonDict['merchantId'] = '';
                
                // Create a ConsultPayRequest object from the JSON request data
                $consultPayRequestObj = ObjectSerializer::deserialize(
                    $jsonDict,
                    'Dana\PaymentGateway\v1\Model\ConsultPayRequest'
                );

            try {                
                // Make the API call - this should fail due to invalid field format
                self::$apiInstance->consultPay($consultPayRequestObj);

            } catch (ApiException $e) {

                // Expected to fail, check if error matches expected response
                Assertion::assertFailResponse(
                    self::$jsonPathFile,
                    self::$titleCase,
                    $caseName,
                    $e->getResponseBody()
                );
            }
        });
    }

    /**
     * Should give fail response code and message for missing mandatory field
     */
    public function testConsultPayInvalidMandatoryField(): void
    {
        try {
            Util::withDelay(function() {
                $caseName = 'ConsultPayBalancedInvalidMandatoryField';

                // Get the request data from the JSON file
                $requestData = Util::getRequest(
                    self::$jsonPathFile,
                    self::$titleCase,
                    $caseName
                );
                
                // Set merchant ID
                $requestData['merchantId'] = self::$merchantId;
                
                // Create headers without timestamp to test validation
                $headers = Util::getHeadersWithSignature(
                    'POST',
                    '/v1.0/payment-gateway/consult-pay.htm',
                    $requestData,
                    false,  // withTimestamp = false to trigger mandatory field error
                    false   // invalidSignature = false
                );

                // Use the executeApiRequest helper method to make manual API call
                try {
                    Util::executeApiRequest(
                        'POST',
                        'https://api.sandbox.dana.id/v1.0/payment-gateway/consult-pay.htm',
                        $headers,
                        $requestData
                    );

                    $this->fail('Expected ApiException for missing X-TIMESTAMP but the API call succeeded');
                } catch (ApiException $e) {
                    // We expect a 400 Bad Request for missing mandatory field
                    $this->assertEquals(400, $e->getCode(), "Expected HTTP 400 Bad Request for missing X-TIMESTAMP, got {$e->getCode()}");

                    // Get the response body from the exception
                    $responseContent = (string)$e->getResponseBody();

                    // Use assertFailResponse to validate the error response
                    Assertion::assertFailResponse(
                        self::$jsonPathFile,
                        self::$titleCase,
                        $caseName,
                        $responseContent
                    );
                } catch (Exception $e) {
                    throw $e;
                }
            });
        } catch (Exception $e) {
            $this->fail("Unexpected exception: " . $e->getMessage());
        }
    }
}
